tests/system/src/test/java/test/javafx/scene/DirtyNodesLeakTest.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package test.javafx.scene;
+
+import javafx.application.Platform;
+import javafx.scene.Scene;
+import javafx.scene.control.Label;
+import javafx.scene.layout.StackPane;
+import javafx.stage.Stage;
+import junit.framework.Assert;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import test.util.Util;
+import test.util.memory.JMemoryBuddy;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+import static test.util.Util.TIMEOUT;
+
+public class DirtyNodesLeakTest {
+
+    @BeforeClass
+    public static void initFX() throws Exception {
+        CountDownLatch startupLatch = new CountDownLatch(1);
+        Platform.setImplicitExit(false);
+        Platform.startup(startupLatch::countDown);
+        Assert.assertTrue("Timeout waiting for FX runtime to start",
+                startupLatch.await(TIMEOUT, TimeUnit.MILLISECONDS));
+    }
+
+    @Test
+    public void leakDirtyNodeAndParentRemoved() throws Exception {
+        JMemoryBuddy.memoryTest((checker) -> {
+            CountDownLatch showingLatch = new CountDownLatch(1);
+            Util.runAndWait(() -> {
+                Stage stage = new Stage();
+                Label label = new Label("Hello!!");
+                StackPane root = new StackPane(label);
+                Scene scene = new Scene(root);
+                stage.setScene(scene);
+
+                checker.setAsReferenced(scene);
+                checker.assertCollectable(label);
+                stage.setOnShown(l -> {
+                    Platform.runLater(() -> {
+                        root.getChildren().clear();
+                        stage.close();
+                        showingLatch.countDown();
+                    });
+                });
```

Comment:
I've been looking at this solution, and from what I see, we're doing our best to add a clean up listener which (as far as I can tell without any comments being added) is called when the stage is no longer shown.

The test above creates a stage, adds a label to it in a stack pane, then clears the children.  I think that should be enough to make the label collectable.

I get the impression that you also have to close the stage to make it collectable, which should not be needed at all.

If I'm correct, then this solution introduces quite some new code but doesn't properly address the issue. The dirty nodes list is clearly holding on to nodes that may not be part of the scene any more. 

Looking at the code, I see that `dirtyNodes` is a poor copy of what `ArrayList` does.  It doesn't clear the array with `null`s when it is resizing the array or when it removes elements from the array (or clears it).

The code that I think is the culprit is this part:

                // This is not the first time this scene has been synchronized,
                // so we will only synchronize those nodes that need it
                for (int i = 0 ; i < dirtyNodesSize; ++i) {
                    Node node = dirtyNodes[i];
                    dirtyNodes[i] = null;
                    if (node.getScene() == Scene.this) {
                            node.syncPeer();
                    }
                }
                dirtyNodesSize = 0;

The problem here is that it sets `dirtyNodesSize` to 0, but doesn't clear the array, leaving references to all kinds of nodes in there.

My recommendation to solve this would be to change this code to use `ArrayList` instead of a hand rolled solution. `ArrayList` will properly `null` out slots in its array when they become unused.

Reply:
> The test above creates a stage, adds a label to it in a stack pane, then clears the children. I think that should be enough to make the label collectable.
Thats correct. In this case it is collectable.

But the point of the test is, to make sure it's also collectable when the stage no longer renders.

I've just looked at it - and to my impression, everything works as expected. 
It's collectable when the stage is showing.
But not collectable when the stage gets immediately closed (which is the bug of this PR)